File: NET/JNetProto.Sharp/Beans/ProtoConvert.cs
using System;
using System.Linq;
using JNetProto.Sharp.API;
using JNetProto.Sharp.Core;
using JNetProto.Sharp.Tools;
using Stream = System.IO.Stream;
using MemoryStream = System.IO.MemoryStream;

namespace JNetProto.Sharp.Beans
{
    public sealed class ProtoConvert : IDisposable
    {
        private readonly object _writeLock = new();
        private readonly object _readLock = new();

        private readonly IDataReader _reader;
        private readonly IDataWriter _writer;
        private readonly ProtoSettings _cfg;

        public ProtoConvert(Stream stdOutput, Stream stdInput, ProtoSettings cfg)
        {
            _reader = new BinaryReader(stdOutput);
            _writer = new BinaryWriter(stdInput);
            _cfg = cfg;
        }

        public void WriteObject(object obj)
        {
            var bytes = SerializeObject(obj, _cfg);
            lock (_writeLock)
            {
                _writer.WriteBinary(bytes);
            }
        }

        public T ReadObject<T>()
        {
            byte[] bytes;
            lock (_readLock)
            {
                bytes = _reader.ReadBinary();
            }
            return DeserializeObject<T>(bytes, _cfg);
        }

        public void Flush()
        {
            lock (_writeLock)
            {
                _writer.Flush();
            }
        }

        public void Dispose()
        {
            _reader?.Dispose();
            _writer?.Dispose();
        }

        private static byte[] SerializeObject(object obj, ProtoSettings s)
        {
            var raw = Conversions.ToObjectArray(obj);
            var args = (object[])raw;
            return SerializeObject(args, s);
        }

        private static T DeserializeObject<T>(object[] args, ProtoSettings _)
        {
            var type = typeof(T);
            var raw = Conversions.FromObjectArray(type, args);
            return (T)raw;
        }

        private static byte[] SerializeObject(object[] args, ProtoSettings _)
        {
            using var mem = new MemoryStream();
            using IDataWriter writer = new BinaryWriter(mem);
            writer.WriteObject(args);
            return mem.ToArray();
        }

        private static T DeserializeObject<T>(byte[] bytes, ProtoSettings s)
        {
            using var mem = new MemoryStream(bytes);
            using IDataReader reader = new BinaryReader(mem);
            var args = (object[])reader.ReadObject();
            return DeserializeObject<T>(args, s);
        }
    }
}

File: JVM/jnetproto-java/src/main/java/jnetproto/java/beans/ProtoConvert.java
package jnetproto.java.beans;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.lang.reflect.Type;

import jnetbase.java.meta.TypeToken;
import jnetproto.java.api.IDataReader;
import jnetproto.java.api.IDataWriter;
import jnetproto.java.core.BinaryReader;
import jnetproto.java.core.BinaryWriter;
import jnetproto.java.tools.Conversions;

public final class ProtoConvert implements AutoCloseable {

    private final Object _writeLock = new Object();
    private final Object _readLock = new Object();

    private final IDataReader _reader;
    private final IDataWriter _writer;
    private final ProtoSettings _cfg;

    public ProtoConvert(InputStream stdOutput, OutputStream stdInput, ProtoSettings cfg) {
        _reader = new BinaryReader(stdOutput);
        _writer = new BinaryWriter(stdInput);
        _cfg = cfg;
    }

    public void writeObject(Object obj) throws Exception {
        byte[] bytes = serializeObject(obj, _cfg);
        synchronized (_writeLock) {
            _writer.writeBinary(bytes);
        }
    }

    public <T> T readObject(Class<T> clazz) throws Exception {
        TypeToken<T> token = TypeToken.wrap(clazz);
        return readObject(token);
    }

    public <T> T readObject(TypeToken<T> type) throws Exception {
        byte[] bytes;
        synchronized (_readLock) {
            bytes = _reader.readBinary();
        }
        return deserializeObject(type, bytes, _cfg);
    }

    public void flush() throws IOException {
        synchronized (_writeLock) {
            _writer.flush();
        }
    }

    @Override
    public void close() throws Exception {
        _reader.close();
        _writer.close();
    }

    private static byte[] serializeObject(Object obj, ProtoSettings s)
            throws Exception {
        Object raw = Conversions.toObjectArray(obj);
        Object[] args = (Object[]) raw;
        return serializeObject(args, s);
    }

    private static <T> T deserializeObject(TypeToken<T> token, Object[] args, ProtoSettings s)
            throws Exception {
        Type type = token.toType();
        Object raw = Conversions.fromObjectArray(type, args);
        return (T) raw;
    }

    private static byte[] serializeObject(Object[] args, ProtoSettings s)
            throws Exception {
        try (ByteArrayOutputStream mem = new ByteArrayOutputStream();
             IDataWriter writer = new BinaryWriter(mem)) {
            writer.writeObject(args);
            return mem.toByteArray();
        }
    }

    private static <T> T deserializeObject(TypeToken<T> clazz, byte[] bytes, ProtoSettings s)
            throws Exception {
        try (ByteArrayInputStream mem = new ByteArrayInputStream(bytes);
             IDataReader reader = new BinaryReader(mem)) {
            Object[] args = (Object[]) reader.readObject();
            return deserializeObject(clazz, args, s);
        }
    }
}
